kernel/src/drivers/keyboard/mod.rs

```diff
@@ -0,0 +1,200 @@
+//! # Keyboard Driver
+//!
+//! The keyboard driver handles all keyboard related functionality, intended to support both PS/2 and USB.
+//! Currently, only PS/2 support has been implemented through the use of the PS/2 driver.
+//!
+//! The driver is event based, and events are received through the `read_event` method, which blocks until an event is received.
+//! The event contains the keycode pressed, which can be compared to `keymap::codes`, an optional `char`, the type of press, and various modifier flags.
+
+pub mod keymap;
+
+use core::convert::From;
+
+use drivers::ps2::{self, Device, DeviceState};
+use drivers::ps2::io::Ps2Error;
+use drivers::ps2::io::commands::{DeviceCommand, DeviceDataCommand};
+
+/// Interface to a generic keyboard.
+pub trait Keyboard {
+    type Error;
+
+    /// Enables this keyboard, setting it up before use.
+    ///
+    /// # Note
+    ///
+    /// The keyboard should not be accessed while not enabled!
+    fn enable(&mut self) -> Result<(), Self::Error>;
+
+    /// Disables this keyboard, making use unavailable.
+    ///
+    /// # Note
+    ///
+    /// Until `enable` is called again, this keyboard should not be used
+    fn disable(&mut self) -> Result<(), Self::Error>;
+
+    /// Polls the device for state events.
+    /// TODO: This should eventually use interrupts and hold a queue
+    fn read_event(&mut self) -> Result<KeyEvent, Self::Error>;
```

Comment:
There could be a `poll_event` which could return how many events are waiting to be read. They could be held in an internal array.